float32 lookahead
float32 horizontal_error